pkg/logs/patterns/tags/tag_manager.go

```diff
@@ -0,0 +1,140 @@
+type TagManager struct {
+	stringToID map[string]uint64
+	nextID     atomic.Uint64
+	mu         sync.RWMutex // prevent other log pipelines from accessing the dictionary concurrently
+}
+
+// NewTagManager creates a new TagManager instance
+func NewTagManager() *TagManager {
+	tm := &TagManager{
+		stringToID: make(map[string]uint64),
+	}
+	return tm
+}
+
+// AddString adds a string to the dictionary and returns its dictionary ID
+// If the string already exists, returns the existing ID
+// Returns the dictionary ID and a boolean indicating if it was newly added
+func (tm *TagManager) AddString(s string) (dictID uint64, isNew bool) {
+	tm.mu.RLock()
+	if id, exists := tm.stringToID[s]; exists {
+		tm.mu.RUnlock()
+		return id, false
+	}
+	tm.mu.RUnlock()
+
+	tm.mu.Lock()
+	defer tm.mu.Unlock()
+	// Double-check locking in case another goroutine added it between locks
+	if id, exists := tm.stringToID[s]; exists {
+		return id, false
+	}
+	// Assign new ID (IDs start at 1)
+	id := tm.nextID.Add(1)
+	tm.stringToID[s] = id
+	return id, true
+}
+
+// EncodeTagStrings converts a slice of "key:value" tag strings into Tag proto messages
+// backed by dictionary indices. It returns the encoded tags plus the dictionary entries
+// that must be flushed upstream (ID -> string) for any newly-seen key/value strings.
+func (tm *TagManager) EncodeTagStrings(tagStrings []string) (tag []*statefulpb.Tag, dict map[uint64]string) {
+	if len(tagStrings) == 0 {
+		return []*statefulpb.Tag{}, map[uint64]string{}
+	}
+
+	encoded := make([]*statefulpb.Tag, 0, len(tagStrings))
+	newEntries := map[uint64]string{}
+
+	for _, tagStr := range tagStrings {
+		if tagStr == "" {
+			continue
+		}
+
+		key, value, hasDelimiter := strings.Cut(tagStr, ":")
+
+		switch {
+		case !hasDelimiter:
+			// Treat bare value tags as key-only tags.
+			keyID, keyNew := tm.AddString(tagStr)
+			if keyNew {
+				newEntries[keyID] = tagStr
+			}
+			encoded = append(encoded, &statefulpb.Tag{
+				Key: dictIndexValue(keyID),
+			})
+		case key == "":
+			// Assume that user mistype the tag string, skip it.
+			log.Warnf("Invalid tag string: %s", tagStr)
+			continue
+		default:
+			keyID, keyNew := tm.AddString(key)
+			if keyNew {
+				newEntries[keyID] = key
+			}
+
+			valueID, valueNew := tm.AddString(value)
+			if valueNew {
+				newEntries[valueID] = value
+			}
+
+			encoded = append(encoded, &statefulpb.Tag{
+				Key:   dictIndexValue(keyID),
+				Value: dictIndexValue(valueID),
+			})
+		}
+	}
+
+	return encoded, newEntries
+}
+
+// GetStringID returns the dictionary ID for a string, if it exists
+// Returns the ID and a boolean indicating if the string was found
+func (tm *TagManager) GetStringID(s string) (uint64, bool) {
+	tm.mu.RLock()
+	defer tm.mu.RUnlock()
+	id, exists := tm.stringToID[s]
+	return id, exists
+}
+
+// Get returns the dictionary ID for a tag string, if it exists
+// Returns the ID and a boolean indicating if the tag was found
+// Deprecated: Use GetStringID instead
```

Comment:
If this is new functionality is there really any value in checking in a deprecated method?